### PR TITLE
CR-1132341 - Update to aie_pl_test to support aie2 on v70

### PR DESCRIPTION
**Pull request by jeffli-xilinx**
Added support for aie2 validate test for v70
Addresses CR-1132341

#### Problem solved by the commit
Enable validate_aie2_pl test for v70

#### Bug / issue (if any) fixed, which PR introduced the bug, how it was discovered
New test 

#### How problem was solved, alternative solutions (if any) and why they were rejected
Added new pl controller sw code for aie2

#### Risks (if any) associated the changes in the commit
None

#### What has been tested and how, request additional testing if necessary
vck5000 and v70

### CODE REVIEW

File: tests/validate/aie_pl_test/src/enums.hpp [hunk 1]
```diff
@@ -1,47 +1,96 @@
...
+  HALT = 9,
+  SET_AIE_ITERATION = 10,
+  ENABLE_AIE_CORES = 11,
+  DISABLE_AIE_CORES = 12,
+  SYNC = 13,
+  LOOP_BEGIN = 14,
+  LOOP_END = 15,
+  SET_DMA_BD = 16,
+  UPDATE_AIE_RTP = 17,
+  WRITE
```
Comment (rozumx) [inline tests/validate/aie_pl_test/src/enums.hpp:88]: Why no value?

File: tests/validate/aie_pl_test/src/host.cpp [hunk 2]
```diff
@@ -148,7 +84,7 @@ main(int argc, char* argv[])
     auto host_out1 = out_bo1.map<int*>();
 
     // input memory
-    auto in_bo1 = xrt::bo(device, mem_size, sender_receiver_k1.group_id(2));
+    auto in_bo1 = xrt::bo(device, mem_size, sender_receiver_k1.group_id(1));
```
Comment (rozumx) [inline tests/validate/aie_pl_test/src/host.cpp:87]: Magic number.  What is so special about 1 and 2?

File: tests/validate/aie_pl_test/src/host.cpp [hunk 3]
```diff
@@ -198,13 +134,203 @@ main(int argc, char* argv[])
     out_bo1.sync(XCL_BO_SYNC_BO_FROM_DEVICE, mem_size, /*OFFSET=*/0);
 
     // post-processing data;
-    bool match = 0;
+    bool match = false;
     for (unsigned int i = 0; i < mem_size / sizeof(int); i++) {
         if (*(host_out1 + i) != *(host_in1 + i) + 1) {
-            match = 1;
+            match = true;
```
Comment (rozumx) [inline tests/validate/aie_pl_test/src/host.cpp:140]: Nice catch

File: tests/validate/aie_pl_test/src/host.cpp [hunk 4]
```diff
@@ -198,13 +134,203 @@ main(int argc, char* argv[])
...
         }
     }
+    return match;
+}
+
+bool run_pl_controller_aie2(xrt::device device, xrt::uuid uuid, std::string aie_control, std::string dma_lock) {
+    // instance of plController
+    xf::plctrl::plController_aie2 m_pl_ctrl(aie_control.c_str(), dma_lock.c_str());
+
+    int num_iter = 1;
```
Comment (rozumx) [inline tests/validate/aie_pl_test/src/host.cpp:151]: Can these values be unsigned int?

File: tests/validate/aie_pl_test/src/host.cpp [hunk 5]
```diff
@@ -198,13 +134,203 @@ main(int argc, char* argv[])
...
+    int num_sample = 32;
+
+    m_pl_ctrl.enqueue_set_aie_iteration("mygraph", num_iter);
+    m_pl_ctrl.enqueue_enable_aie_cores();
+
+    for (int i = 0; i < num_iter; ++i) {
```
Comment (rozumx) [inline tests/validate/aie_pl_test/src/host.cpp:157]: Curly brace is not needed.

File: tests/validate/aie_pl_test/src/host.cpp [hunk 6]
```diff
@@ -198,13 +134,203 @@ main(int argc, char* argv[])
...
+        m_pl_ctrl.enqueue_sync();
+    }
+
+    m_pl_ctrl.enqueue_sleep(128);
```
Comment (rozumx) [inline tests/validate/aie_pl_test/src/host.cpp:161]: Magic number

File: tests/validate/aie_pl_test/src/host.cpp [hunk 7]
```diff
@@ -198,13 +134,203 @@ main(int argc, char* argv[])
...
+    m_pl_ctrl.enqueue_disable_aie_cores();
+
+    m_pl_ctrl.enqueue_halt();
+    m_pl_ctrl.print_micro_codes();
+
+    int ret;
```
Comment (rozumx) [inline tests/validate/aie_pl_test/src/host.cpp:167]: Uninitialized variable

File: tests/validate/aie_pl_test/src/host.cpp [hunk 8]
```diff
@@ -198,13 +134,203 @@ main(int argc, char* argv[])
...
+
+    // XRT auto get group_id
+    auto sender_receiver_k1 =
+        xrt::kernel(device, uuid, "sender_receiver:{sender_receiver_1}");
+    auto controller_k1 =
+        xrt::kernel(device, uuid, "pl_controller_top:{controller_1}");
+
+    // output memory
+    mem_size = num_sample * num_iter * sizeof(int);
+    auto out_bo1 = xrt::bo(device, mem_size, sender_receiver_k1.group_id(3));
```
Comment (rozumx) [inline tests/validate/aie_pl_test/src/host.cpp:179]: Magic number

File: tests/validate/aie_pl_test/src/host.cpp [hunk 9]
```diff
@@ -198,13 +134,203 @@ main(int argc, char* argv[])
...
+    auto host_out1 = out_bo1.map<int*>();
+
+    // input memory
+    auto in_bo1 = xrt::bo(device, mem_size, sender_receiver_k1.group_id(2));
```
Comment (rozumx) [inline tests/validate/aie_pl_test/src/host.cpp:183]: Magic number

File: tests/validate/aie_pl_test/src/host.cpp [hunk 10]
```diff
@@ -198,13 +134,203 @@ main(int argc, char* argv[])
...
+    auto host_in1 = in_bo1.map<int*>();
+    std::cout << " memory allocation complete" << std::endl;
+
+    // initialize input memory
+    for (int i = 0; i < mem_size / sizeof(int); i++) {
```
Comment (rozumx) [inline tests/validate/aie_pl_test/src/host.cpp:188]: Curly braces are not needed

File: tests/validate/aie_pl_test/src/host.cpp [hunk 11]
```diff
@@ -198,13 +134,203 @@ main(int argc, char* argv[])
...
+    int mem_size = 0;
+
+    // XRT auto get group_id
+    auto sender_receiver_k1 =
+        xrt::kernel(device, uuid, "sender_receiver:{sender_receiver_1}");
+    auto controller_k1 =
+        xrt::kernel(device, uuid, "pl_controller_top:{controller_1}");
+
+    // output memory
+    mem_size = num_sample * num_iter * sizeof(int);
```
Comment (rozumx) [inline tests/validate/aie_pl_test/src/host.cpp:178]: Do you really wish to use sizeof(int)?  I ask for other portions of this method uses sizeof(uint32_t)

File: tests/validate/aie_pl_test/src/host.cpp [hunk 12]
```diff
@@ -198,13 +134,203 @@ main(int argc, char* argv[])
...
+    controller_r1.set_arg(3, pm_bo);
+    controller_r1.start();
+    std::cout << "start pl controller kernel" << std::endl;
+    // start input kernels
+
+    controller_r1.wait();
+    // sync output memory
+    out_bo1.sync(XCL_BO_SYNC_BO_FROM_DEVICE, mem_size, /*OFFSET=*/0);
+    // post-processing data;
+    int i;
```
Comment (rozumx) [inline tests/validate/aie_pl_test/src/host.cpp:229]: This variable and be used in the for statement.  It should be remove.
```suggestion

```

File: tests/validate/aie_pl_test/src/host.cpp [hunk 13]
```diff
@@ -198,13 +134,203 @@ main(int argc, char* argv[])
...
+
+int
+main(int argc, char* argv[])
+{
+  std::string b_file;
+
+    // Option Variables
+    std::string test_path;
+    std::string dev_id = "0";
+    bool flag_s = false;
```
Comment (rozumx) [inline tests/validate/aie_pl_test/src/host.cpp:248]: Please rename to a more useful variable name.  For example: bSupported

File: tests/validate/aie_pl_test/src/host.cpp [hunk 14]
```diff
@@ -198,13 +134,203 @@ main(int argc, char* argv[])
...
+        std::cerr << "ERROR: " << e.what() << std::endl << std::endl;
+        std::cout << desc << "\n";
+        return EXIT_FAILURE;
+    }
+
+    std::string aie_control_file = "/aie_control_config.json";
+    std::string aie_control = test_path + aie_control_file;
+    std::ifstream aiefile(aie_control);
+    if (!aiefile.good()) {
+        std::cout << "\nNOT SUPPORTED" << std::endl;
```
Comment (rozumx) [inline tests/validate/aie_pl_test/src/host.cpp:291]: Should this be an error and sent out to std::cerr

File: tests/validate/aie_pl_test/src/host.cpp [hunk 15]
```diff
@@ -198,13 +134,203 @@ main(int argc, char* argv[])
...
+        return EOPNOTSUPP;
+    }
+
+    ptree aie_meta;
+    read_json(aie_control, aie_meta);
+    auto driver_info_node = aie_meta.get_child("aie_metadata.driver_config");
+    auto hw_gen_node = driver_info_node.get_child("hw_gen");
+    auto hw_gen = std::stoul(hw_gen_node.data());
+
+    if(hw_gen == 2) {
```
Comment (rozumx) [inline tests/validate/aie_pl_test/src/host.cpp:301]: Magic number and curly braces are not needed.

File: tests/validate/aie_pl_test/src/host.cpp [hunk 16]
```diff
@@ -198,13 +134,203 @@ main(int argc, char* argv[])
...
+      b_file  = "/v70_pl_controller.xclbin";
+    } else {
+      b_file  = "/vck5000_pcie_pl_controller.xclbin";
+    }
+    
+    std::string binaryfile = test_path + b_file;
```
Comment (rozumx) [inline tests/validate/aie_pl_test/src/host.cpp:307]: Use boost::filesystem::path to join paths and files.

File: tests/validate/aie_pl_test/src/host.cpp [hunk 17]
```diff
@@ -198,13 +134,203 @@ main(int argc, char* argv[])
...
+    std::ifstream infile(binaryfile);
+
+    if (!infile.good()) {
+        std::cout << "\nNOT SUPPORTED" << std::endl;
```
Comment (rozumx) [inline tests/validate/aie_pl_test/src/host.cpp:311]: Shouldn't this be an error?

File: tests/validate/aie_pl_test/src/host.cpp [hunk 18]
```diff
@@ -198,13 +134,203 @@ main(int argc, char* argv[])
...
+        return EXIT_SUCCESS;
+    }
+
+    auto num_devices = xrt::system::enumerate_devices();
+    auto device = xrt::device{ dev_id };
+    auto uuid = device.load_xclbin(binaryfile);
+
+    // instance of plController
+    std::string dma_lock_file = "/dma_lock_report.json";
+    std::string dma_lock = test_path + dma_lock_file;
```
Comment (rozumx) [inline tests/validate/aie_pl_test/src/host.cpp:326]: Use boost::filesystem::path

File: tests/validate/aie_pl_test/src/host.cpp [hunk 19]
```diff
@@ -198,13 +134,203 @@ main(int argc, char* argv[])
...
+
+    bool match = false;
+    if (hw_gen == 2) {
```
Comment (rozumx) [inline tests/validate/aie_pl_test/src/host.cpp:329]: Magic number

File: tests/validate/aie_pl_test/src/pl_controller_aie2.hpp [hunk 20]
```diff
@@ -0,0 +1,161 @@
...
+    dynBuffer() {
+        data = nullptr;
+        size = 0;
+        usedSize = 0;
+    }
+    uint32_t get(int i) {
+        assert(i < usedSize);
+        return data[i];
+    }
+    void add(uint32_t m_data) {
```
Comment (rozumx) [inline tests/validate/aie_pl_test/src/pl_controller_aie2.hpp:52]: Is there a reason why all of these method bodies are in the header file and not body file?

File: tests/validate/aie_pl_test/src/pl_controller_aie2.hpp [hunk 21]
```diff
@@ -0,0 +1,161 @@
...
+        if (size == 0) {
+            size = 128;
```
Comment (rozumx) [inline tests/validate/aie_pl_test/src/pl_controller_aie2.hpp:54]: Magic number.

**Review by rozumx — CHANGES_REQUESTED** [review]

File: tests/validate/aie_pl_test/src/enums.hpp (hunk 1 above)
Comment (jeffli-xilinx) [reply tests/validate/aie_pl_test/src/enums.hpp:88]: This was code from aie2 controller code in Vitis example designs.
I'll just update it to be 18 to be explicit.

File: tests/validate/aie_pl_test/src/host.cpp (hunk 14 above)
Comment (jeffli-xilinx) [reply tests/validate/aie_pl_test/src/host.cpp:291]: This is how all the other validate tests are handled.  
I think xbutil will take the string output and decide if the test can be supported or not.

File: tests/validate/aie_pl_test/src/pl_controller_aie2.hpp (hunk 20 above)
Comment (jeffli-xilinx) [reply tests/validate/aie_pl_test/src/pl_controller_aie2.hpp:52]: I can move these into the body file.  I just took the example code and dropped it in.

**Comment by dayeh-xilinx** [comment]
retest this please.

**Comment by dayeh-xilinx** [comment]
retest this please.

**Comment by dayeh-xilinx** [comment]
retest this please.

File: tests/validate/aie_pl_test/src/pl_controller_aie2.hpp [hunk 22]
```diff
@@ -0,0 +1,131 @@
...
+    int read_elf_to_mem(std::string file_name, dynBuffer* m_buff);
+    // re-use this code from "core/edge/common/aie_parser.cpp"
+    void get_rtp();
+
+    std::vector<tile_type> get_tiles(const std::string& graph_name);
+
+    std::vector<buffer_type> get_buffers(const std::string& port_name);
+
+    std::vector<char> m_axlf;
+    // const axlf* m_top;
```
Comment (rozumx) [inline tests/validate/aie_pl_test/src/pl_controller_aie2.hpp:117]: Dead code

File: tests/validate/aie_pl_test/src/pl_controller_aie2.hpp [hunk 23]
```diff
@@ -0,0 +1,131 @@
...
+    std::unordered_map<std::string, rtp_type> rtps;
+    dynBuffer* opcodeBuffer;
```
Comment (rozumx) [inline tests/validate/aie_pl_test/src/pl_controller_aie2.hpp:119]: Why does this need to be a pointer.

Comment (rozumx) [inline tests/validate/aie_pl_test/src/pl_controller_aie2.hpp:119]: Member variables need to have a "m_" prefix.

File: tests/validate/aie_pl_test/src/pl_controller_aie2.hpp [hunk 24]
```diff
@@ -0,0 +1,131 @@
...
+    uint32_t get(int i);
+    void add(uint32_t m_data);
+    void add(uint32_t* m_data, int blk_size);
+};
+
+class plController_aie2 {
+   public:
+    /* Constructor
+    */
+    plController_aie2(const std::string& xclbin_path);
```
Comment (rozumx) [inline tests/validate/aie_pl_test/src/pl_controller_aie2.hpp:55]: What are you doing with regards to the default constructor?

Comment (rozumx) [inline tests/validate/aie_pl_test/src/pl_controller_aie2.hpp:55]: Isn't this constructor the same as the one on line 56 with the the dma_info_path being empty?

File: tests/validate/aie_pl_test/src/pl_controller_aie2.cpp [hunk 25]
```diff
@@ -0,0 +1,426 @@
...
+                printf("UPDATE_AIE_RTP: rtp_val=%d, addr=%di, ctrl_strm_id=%d\n", rtp_val, addr, ctrl_strm_id);
+                break;
+            }
+            defalut:
+                printf("Not supported opcode %d\n", op);
+        }
+    }
+}
+void plController_aie2::enqueue_set_aie_iteration(const std::string& graphName, int num_iter, int ctrl_nm) {
+    if (num_iter < 0) throw std::runtime_error("Number of iteration < 0: " + std::to_string(num_iter));
```
Comment (rozumx) [inline tests/validate/aie_pl_test/src/pl_controller_aie2.cpp:160]:
```suggestion
    if (num_iter < 0) 
        throw std::runtime_error("Number of iteration < 0: " + std::to_string(num_iter));
```

**Review by rozumx — CHANGES_REQUESTED** [review]
The majority of the issue discovered during the code review have to do with attention to details.  This includes but is not limited to:

1. Copyright dates
2. Formatting.  This includes:
- Readability of the code
- Code on one line that should be split to two (e.g., if statements)
- Code on two lines that should be merged to one
- Function arguments that shouldn't have the underscore prefix '_'
4. Member variables missing the "m_" prefix
5. Class member variables not be initialized correctly in the constructors
6. Extra constructors
7. Enumeration values not being scoped correct
8. Excessive use of using namespace

File: tests/validate/aie_pl_test/src/host.cpp [hunk 26]
```diff
@@ -198,13 +137,203 @@ main(int argc, char* argv[])
...
+
+    // input/output memory for pl_controller
+    in_bo1.sync(XCL_BO_SYNC_BO_TO_DEVICE, mem_size, /*OFFSET=*/0);
+
+    int32_t num_pm = m_pl_ctrl.get_microcode_size(); /// sizeof(int32_t);
+    auto pm_bo = xrt::bo(device, (num_pm + 1) * sizeof(uint32_t),
+                         controller_k1.group_id(3));
+    auto host_pm = pm_bo.map<uint32_t*>();
+
+    m_pl_ctrl.copy_to_device_buff(host_pm + 1);
```
[the remark this answers is not in the document]
Comment (jeffli-xilinx) [reply tests/validate/aie_pl_test/src/host.cpp:204]: I believe the reason is that first entry in the buffer is the number of program memory lines, followed by the program instructions.

File: src/runtime_src/core/tools/xbutil2/SubCmdValidate.cpp [hunk 27]
```diff
@@ -279,7 +279,8 @@ runTestCase( const std::shared_ptr<xrt_core::device>& _dev, const std::string& p
       { "versal_23_bandwidth.py",   "kernel_bw.exe"   },
```
Comment (rozumx) [inline src/runtime_src/core/tools/xbutil2/SubCmdValidate.cpp:279]: Copyright date on line 2 needs to be updated.

**Review by rozumx — APPROVED** [review]

**Comment by dayeh-xilinx** [comment]
retest this please.

**Review by chvamshi-xilinx — APPROVED** [review]

**Pull request merged**